### ISSUE TITLE
when config.user was wrong, always got an uncaught Exception  

### ISSUE THREAD

**Issue by runfan** [mscdex/node-imap#532]
```log
uncaughtException:  { [Error: LOGIN failed] type: 'no', textCode: undefined, source: 'authentication' }
Error: LOGIN failed
    at Connection._resTagged (D:\workspace\com.shaozi.client\Project\ShaoZi.PC\deploy\package.nw\node_modules\imap\lib\Connection.js:1488:11)
    at Parser.<anonymous> (D:\workspace\com.shaozi.client\Project\ShaoZi.PC\deploy\package.nw\node_modules\imap\lib\Connection.js:193:10)
    at emitOne (events.js:75:13)
    at Parser.emit (events.js:150:7)
    at Parser._resTagged (D:\workspace\com.shaozi.client\Project\ShaoZi.PC\deploy\package.nw\node_modules\imap\lib\Parser.js:175:10)
    at Parser._parse (D:\workspace\com.shaozi.client\Project\ShaoZi.PC\deploy\package.nw\node_modules\imap\lib\Parser.js:139:16)
    at Parser._tryread (D:\workspace\com.shaozi.client\Project\ShaoZi.PC\deploy\package.nw\node_modules\imap\lib\Parser.js:82:15)
    at Socket.Parser._cbReadable (D:\workspace\com.shaozi.client\Project\ShaoZi.PC\deploy\package.nw\node_modules\imap\lib\Parser.js:53:12)
    at emitNone (events.js:65:13)
    at Socket.emit (events.js:147:7)
    at emitReadable_ (_stream_readable.js:408:10)
    at emitReadable (_stream_readable.js:402:7)
    at readableAddChunk (_stream_readable.js:157:11)
    at Socket.Readable.push (_stream_readable.js:109:10)
    at TCP.onread (net.js:510:20)
```

**Comment by runfan** [comment]
```Javascript
return new Promise(function (resolve, reject) {
        var $this = this;

        var done = function () {
            logger.debug('done => connect');
            $this._imap.removeListener('error', fail);
            resolve();
        }
        var fail = function (error) {
            if(error){
                logger.debug('error => connect', error);
                $this._imap.removeListener('ready', done);
                reject(error);
            }
        }

        logger.debug('prepare => connect');
        this._imap.once('ready', done);
        this._imap.once('error', fail);
        this._imap.once('end', fail);
        try{
            this._imap.connect();
        }catch(error){
            logger.debug('error => connect', error);
            $this._imap.removeListener('ready', done);
            $this._imap.removeListener('error', fail);
            $this._imap.removeListener('end', fail);
            reject(error);
        }
        logger.debug('begin => connect');
    }.bind(this));
```

**Comment by mscdex** [comment]
What does your logger output? Also, can you set `debug: console.log` in your connection config object and post the output?

**Comment by runfan** [comment]
```log
<= '* OK AliYun IMAP Server Ready(10.177.1.26)'
<= '* CAPABILITY IMAP4rev1 IDLE XLIST UIDPLUS ID SASL-IR AUTH=XOAUTH AUTH=EXTERNAL'
<= 'A0 OK CAPABILITY completed'
=> 'A1 LOGIN "fanchangwang@shzi.com" "xxxx"'
<= 'A1 NO LOGIN failed'
```
need `fanchangwang@shaozi.com`  and i set `fanchangwang@shzi.com`

**Comment by mscdex** [comment]
What does your `logger.debug()` output show?

**Comment by runfan** [comment]
show nothing

**Comment by runfan** [comment]
sorry i did not write it, let me try again.

**Comment by mscdex** [comment]
If you don't see any `logger.debug()` output, what if you replace the `logger.debug(...)` statements with `console.log(...)` instead?

**Comment by runfan** [comment]
```log
[2016-02-25 16:40:02.261] [DEBUG] client - prepare => connect
[2016-02-25 16:40:02.264] [DEBUG] client - begin => connect
[2016-02-25 16:40:02.264] [DEBUG] service - find thread list
[2016-02-25 16:40:02.646] [ERROR] app.js - uncaughtException:  { [Error: LOGIN failed] type: 'no', textCode: undefined, source: 'authentication' }
Error: LOGIN failed
    at Connection._resTagged (D:\workspace\com.shaozi.client\Project\ShaoZi.PC\deploy\package.nw\node_modules\imap\lib\Connection.js:1488:11)
    at Parser.<anonymous> (D:\workspace\com.shaozi.client\Project\ShaoZi.PC\deploy\package.nw\node_modules\imap\lib\Connection.js:193:10)
    at emitOne (events.js:75:13)
    at Parser.emit (events.js:150:7)
    at Parser._resTagged (D:\workspace\com.shaozi.client\Project\ShaoZi.PC\deploy\package.nw\node_modules\imap\lib\Parser.js:175:10)
    at Parser._parse (D:\workspace\com.shaozi.client\Project\ShaoZi.PC\deploy\package.nw\node_modules\imap\lib\Parser.js:139:16)
    at Parser._tryread (D:\workspace\com.shaozi.client\Project\ShaoZi.PC\deploy\package.nw\node_modules\imap\lib\Parser.js:82:15)
    at Socket.Parser._cbReadable (D:\workspace\com.shaozi.client\Project\ShaoZi.PC\deploy\package.nw\node_modules\imap\lib\Parser.js:53:12)
    at emitNone (events.js:65:13)
    at Socket.emit (events.js:147:7)
    at emitReadable_ (_stream_readable.js:408:10)
    at emitReadable (_stream_readable.js:402:7)
    at readableAddChunk (_stream_readable.js:157:11)
    at Socket.Readable.push (_stream_readable.js:109:10)
    at TCP.onread (net.js:510:20)
```

**Comment by runfan** [comment]
i use `node-imap` in `node-webkit v0.12.3`

**Comment by mscdex** [comment]
I don't see any reason why the error shouldn't get caught by your `error` event handler. When I try adding a one-time `error` handler and use an incorrect username, the event handler gets called as expected.

Out of curiosity, have you tried the subset of the code that connects to the imap server with just plain node (a modern version like v4.x or v5.x)?

**Comment by runfan** [comment]
node v4.3.0
```javascript
connect() {
        let p;
        if (!this._imap) {
            p = Promise.reject(new Error('connection is undefined'));
        } else if (this.state === 'authenticated') {
            p = Promise.resolve();
        } else if (this.state === 'connected' || this.state === 'connecting') {
            this._state = 'connectiong';
            p = new Promise((resolve, reject) => {
                if (this._wait < WAIT_FOR_TIMES) {
                    this._wait++;
                    setTimeout(() => {
                        this.connect().then(resolve, reject);
                    }, WAIT_FOR_TIMEOUT);
                } else {
                    let error = new Error('timeout when try to connect');
                    error.code = 'connection timeout';
                    reject(error);
                }
            });
        } else {
            this._state = 'connecting';
            p = new Promise((resolve, reject) => {
                var err = function(error){
                    this._state = false;
                    this._imap.removeListener('ready',done);
                    reject();
                }.bind(this);
                var done = function(){
                    this._state = false;
                    this._imap.removeListener('error',err);
                    resolve();
                }.bind(this)
                this._imap.once('error', err);
                this._imap.once('ready', done);
                this._imap.connect();
            });
        }
        return p;
    }
```
and mocha test  **faild**

```javascript


describe('Connection', function () {
    options.debug = function (msg) {
        console.info(msg);
    };
    let conn = new Connection(new Imap(options));
    describe('#connect()', function () {
        it('try connect', () => conn.connect());
        it('try connect again', () => conn.connect());
    });
});

```

```log
 Connection
    #connect()
[connection] Connected to host
<= '* OK AliYun IMAP Server Ready(10.177.10.32)'
=> 'A0 CAPABILITY'
<= '* CAPABILITY IMAP4rev1 IDLE XLIST UIDPLUS ID SASL-IR AUTH=XOAUTH AUTH=EXTERNAL'
<= 'A0 OK CAPABILITY completed'
=> 'A1 LOGIN "fanchangwang@shzi.com" "Fan200107"'
<= 'A1 NO LOGIN failed'
      1) try connect
[connection] Ended
[connection] Closed
[connection] Connected to host

1) Connection #connect() try connect:
     Error: Promise rejected with no or falsy reason
      at C:\Users\run_f\AppData\Roaming\npm\node_modules\mocha\lib\runnable.js:326:26
      at tryCatcher (d:\workspace\cw.mail\node_modules\bluebird\js\release\util.js:16:23)
      at Promise._settlePromiseFromHandler (d:\workspace\cw.mail\node_modules\bluebird\js\release\promise.js:497:31)
      at Promise._settlePromise (d:\workspace\cw.mail\node_modules\bluebird\js\release\promise.js:555:18)
      at Promise._settlePromise0 (d:\workspace\cw.mail\node_modules\bluebird\js\release\promise.js:600:10)
      at Promise._settlePromises (d:\workspace\cw.mail\node_modules\bluebird\js\release\promise.js:679:18)
      at Async._drainQueue (d:\workspace\cw.mail\node_modules\bluebird\js\release\async.js:125:16)
      at Async._drainQueues (d:\workspace\cw.mail\node_modules\bluebird\js\release\async.js:135:10)
      at Immediate.Async.drainQueues [as _onImmediate] (d:\workspace\cw.mail\node_modules\bluebird\js\release\async.js:16:14)
```

**Comment by runfan** [comment]
my `function connect` catche nothing

**Comment by mscdex** [comment]
That looks to me like it's working then, you're just not passing the error to `reject()` in your `error` event handler.

**Comment by runfan** [comment]
yes you are right , thanks

**Comment by runfan** [comment]
is that means `node-imap` not working in `node-webkit v0.12.3`?

**Comment by mscdex** [comment]
Possibly. You might try with the newer/beta version of node-webkit? It uses node v5 instead of io.js v1.

**Comment by runfan** [comment]
:cry: i can't
